### PR TITLE
Initial support for elliptic build plate shape

### PR DESCRIPTION
**Pull request by fieldOfView**
This PR provides an option to change the build plate shape to elliptic for polar printers.
polar
An elliptic build volume is drawn, and disallowed area polygons are added so that only an elliptic area remains.

### CODE REVIEW

**Comment by nallath** [comment]
You are making a lot of people happy with this! I'll have a look right away.

File: resources/definitions/fdmprinter.def.json [hunk 1]
```diff
@@ -137,6 +137,21 @@
                     "settable_per_extruder": false,
                     "settable_per_meshgroup": false
                 },
+                "machine_shape":
+                {
+                    "label": "Build plate shape",
+                    "description": "The shape of the build plate without taking unprintable areas into account.",
+                    "default_value": "Rectangular",
```
Comment (nallath) [inline resources/definitions/fdmprinter.def.json]: This should be "rectangular" (as per key you used in the options).

**Comment by nallath** [comment]
ellipicalbuildplatefail

File: resources/definitions/fdmprinter.def.json (hunk 1 above)
Comment (fieldOfView) [inline resources/definitions/fdmprinter.def.json]: Good catch. Fixed

File: cura/BuildVolume.py [hunk 2]
```diff
@@ -177,34 +183,75 @@ def rebuild(self):
...
+            self._grid_mesh = mb.build()
+
+        else:
+            # Bottom and top 'ellipse' of the build volume
+            aspect = 1.0
+            scale_matrix = Matrix()
+            if self._width != 0:
+                # Scale circular meshes by aspect ratio if width != height
+                aspect = self._height / self._width
+                scale_matrix.compose(Vector(1, 1, aspect))
```
Comment (fieldOfView) [inline cura/BuildVolume.py]: That's what I meant in the first place. Fixed.

**Comment by fieldOfView** [comment]
Eh? Wt..

If the other half is visible from below, that suggests an error in the winding order of the faces more than an issue with UVs. But the winding order is just fine for me. Will see if I can repro this on linux, or on another machine.

**Comment by fieldOfView** [comment]
Euhm, in your screenshot (probably not related to this issue)... How come Dual Extrusion Overlap is showing in the Build Plate Adhesion category, and why is it showing at all for a single extrusion printer?

**Comment by nallath** [comment]
Euh. Good one. I have no idea why that went wrong.

**Comment by Ghostkeeper** [comment]
Good catch.
Turns out that the dual extrusion category header gets hidden for single extrusion, but we don't disable the two settings Dual Extrusion Overlap and Ooze Shield Enabled for single extrusion. Because there is no header, it appears as if the setting is under the Build Plate Adhesion category. Collapsing the Build Plate Adhesion category doesn't hide the Dual Extrusion Overlap setting.

**Comment by fieldOfView** [comment]
Back to the rendering issue with the elliptic buildplate: it looks like on Linux, the indices I set are being ignored, and instead the faces are always drawn as a triangle strip.

**Comment by fieldOfView** [comment]
I was not adding indices correctly. Took a while to figure this out because Windows was rendering it correctly regardless.

**Comment by nickthetait** [comment]
So people can now use Cura 2 with almost any delta printer right? Are there other features that are still needed?

**Comment by fieldOfView** [comment]
* Handling of unreachable areas due to extruder offsets is not as effective as for rectangular printers
* Auto scaling / scale to max does not work very well on elliptic platforms

There may be lingering issues with center_is_zero in the engine (I know there were issues before, and the feature gets next to 0 testing at UM because their printers don't use that feature)

**Comment by nallath** [comment]
Scale to max doesn't work that well on any printer. So not really a loss there.

**Pull request merged**